subsys/modbus/mb_rtu_core.c

```diff
@@ -0,0 +1,654 @@
+		return -EMSGSIZE;
+	}
+
+	/* Take away for the ':', CR, and LF */
+	rx_size -= 3;
+	/* Point past the ':' to the address. */
+	pmsg = &ctx->uart_buf[1];
+
+	hex2bin(pmsg, 2, &ctx->rx_frame.addr, 1);
+	pmsg += 2;
+	rx_size -= 2;
+	hex2bin(pmsg, 2, &ctx->rx_frame.fc, 1);
+	pmsg += 2;
+	rx_size -= 2;
+
+	/* Get the data from the message */
+	ctx->rx_frame.length = 0;
+	while (rx_size > 2) {
+		hex2bin(pmsg, 2, prx_data, 1);
+		prx_data++;
+		pmsg += 2;
+		rx_size -= 2;
+		/* Increment the number of Modbus packets received */
+		ctx->rx_frame.length++;
+	}
+
+	/* Subtract the Address and function code */
+	ctx->rx_frame.length -= 2;
+	/* Extract the message's LRC */
+	hex2bin(pmsg, 2, &frame_lrc, 1);
+	ctx->rx_frame.crc = frame_lrc;
+
+	/*
+	 * The LRC is calculated on the ADDR, FC and Data fields,
+	 * not the ':', CR/LF and LRC placed in the message
+	 * by the sender. We thus need to subtract 5 'ASCII' characters
+	 * from the received message to exclude these.
+	 */
+	calc_lrc = mb_ascii_get_lrc(&ctx->uart_buf[1],
+				    (ctx->uart_buf_ctr - 5) / 2);
+
+	ctx->uart_buf_ctr = 0;
+	ctx->uart_buf_ptr = &ctx->uart_buf[0];
+
+	if (calc_lrc != frame_lrc) {
+		LOG_ERR("Calculated LRC does not match received LRC");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static uint8_t *mb_bin2hex(uint8_t value, uint8_t *pbuf)
+{
+	uint8_t u_nibble = (value >> 4) & 0x0F;
+	uint8_t l_nibble = value & 0x0F;
+
+	hex2char(u_nibble, pbuf);
+	pbuf++;
+	hex2char(l_nibble, pbuf);
+	pbuf++;
+
+	return pbuf;
+}
+
+static void mb_tx_ascii_frame(struct mb_rtu_context *ctx)
+{
+	uint16_t tx_bytes = 0;
+	uint8_t lrc;
+	uint8_t *pbuf;
+
+	/* Place the start-of-frame character into output buffer */
+	ctx->uart_buf[0] = MODBUS_ASCII_START_FRAME_CHAR;
+	tx_bytes = 1;
+
+	pbuf = &ctx->uart_buf[1];
+	pbuf = mb_bin2hex(ctx->tx_frame.addr, pbuf);
+	tx_bytes += 2;
+	pbuf = mb_bin2hex(ctx->tx_frame.fc, pbuf);
+	tx_bytes += 2;
+
+	for (int i = 0; i < ctx->tx_frame.length; i++) {
+		pbuf = mb_bin2hex(ctx->tx_frame.data[i], pbuf);
+		tx_bytes += 2;
+	}
+
+	/*
+	 * Add the LRC checksum in the packet.
+	 *
+	 * The LRC is calculated on the ADDR, FC and Data fields,
+	 * not the ':' which was inserted in the uart_buf[].
+	 * Thus we subtract 1 ASCII character from the LRC.
+	 * The LRC and CR/LF bytes are not YET in the .uart_buf[].
+	 */
+	lrc = mb_ascii_get_lrc(&ctx->uart_buf[1], (tx_bytes - 1) / 2);
+	pbuf = mb_bin2hex(lrc, pbuf);
+	tx_bytes += 2;
+
+	*pbuf++ = MODBUS_ASCII_END_FRAME_CHAR1;
+	*pbuf++ = MODBUS_ASCII_END_FRAME_CHAR2;
+	tx_bytes += 2;
+
+	/* Update the total number of bytes to send */
+	ctx->uart_buf_ctr = tx_bytes;
+	ctx->uart_buf_ptr = &ctx->uart_buf[0];
+
+	LOG_DBG("Start frame transmission");
+	uart_irq_rx_disable(ctx->dev);
+	uart_irq_tx_enable(ctx->dev);
+}
+#else
+static int mb_rx_ascii_frame(struct mb_rtu_context *ctx)
+{
+	return 0;
+}
+
+static void mb_tx_ascii_frame(struct mb_rtu_context *ctx)
+{
+}
+#endif
+
+static uint16_t mb_rtu_crc16(uint8_t *src, size_t length)
```

Comment:
Is it possible to use Zephyr CRC facilities? https://docs.zephyrproject.org/apidoc/latest/group__crc.html

Reply:
AFAIK no, this kind "reflected" algorithm is not supported by `lib/os/crc16_sw.c`